### PR TITLE
Topic/schedule pipelined

### PR DESCRIPTION
**Pull request by vspetrov**
## What
Adds pipelined schedule interface and the  corresponding implementation for TL/UCP SRA allreduce algorithm

## Why ?
Pipelining improves performance of large message collectives in multiple cases

## How ?
- Pipelined schedule data structure is introduced that runs a collective as a sequence of fragments
- The interface is generic: user (tl/ucp e.g.) passes a cb that allocates a schedule -> it defines an execution of a collective fragment
- Number of simultaneously allocated fragments - pipeline depth - can be varied. 
- Frag schedules are restarted automatically keeping the dependencies
- 2 types of pipelines are supported: parallel and sequential. In parallel pipeline the tasks at the same position in frag schedule can be started out of order, in sequential the order is fixed (needed for example when TL that is used in frag schedule can't support out of order tasks)

### CODE REVIEW

**Comment by vspetrov** [comment]
bot:retest

**Review by manjugv — APPROVED** [review]

File: src/schedule/ucc_schedule_pipelined.c [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    schedule->super.n_tasks        = n_frags_total;
+    schedule->n_frags              = n_frags;
+    schedule->sequential           = sequential;
+    schedule->frag_setup           = frag_setup;
+    schedule->next_frag_to_post    = 0;
+    schedule->n_frags_in_pipeline  = 0;
+    schedule->super.super.finalize = ucc_schedule_pipelined_finalize;
+    schedule->super.super.post     = ucc_schedule_pipelined_post;
+    frags                          = schedule->frags;
+    for (i = 0; i < n_frags; i++) {
```
Comment (Sergei-Lebedev) [inline src/schedule/ucc_schedule_pipelined.c:171]: check n_frags < max_frags, return error othrewise

File: src/schedule/ucc_schedule_pipelined.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    ucc_schedule_t *             frags[UCC_SCHEDULE_PIPELINED_MAX_FRAGS];
+    int                          n_frags;
+    int                          n_frags_started;
+    int                          n_frags_in_pipeline;
+    int                          sequential;
+    int                          next_frag_to_post;
+    ucc_schedule_frag_setup_fn_t frag_setup;
+} ucc_schedule_pipelined_t;
+
+ucc_status_t ucc_schedule_pipelined_init(
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/schedule/ucc_schedule_pipelined.h:33]: done

**Comment by vspetrov** [comment]
@bureddy i've modified "sequential" pipeline to depend on "EVENT_STARTED" instead of completed (according to discussion in WG). @Sergei-Lebedev your comments are also addressed. so, plz review when you have time (and then i can post new hierarchical CL PR).

File: test/gtest/core/test_allreduce.cc [hunk 3]
```diff
@@ -242,3 +242,48 @@ TYPED_TEST(test_allreduce, multiple_cuda_inplace) {
...
+                             {"UCC_TL_UCP_TUNE", "allreduce:@sra_knomial:inf"},
+                             {"UCC_TL_UCP_ALLREDUCE_SRA_KN_FRAG_THRESH", "1024"},
+                             {"UCC_TL_UCP_ALLREDUCE_SRA_KN_N_FRAGS", "11"}};
+    UccJob        job(n_procs, UccJob::UCC_JOB_CTX_GLOBAL, env);
+    UccTeam_h     team   = job.create_team(n_procs);
+    int           repeat = 3;
+    UccCollCtxVec ctxs;
+    std::vector<ucc_memory_type_t> mt = {UCC_MEMORY_TYPE_HOST};
+
+#ifdef HAVE_CUDA
```
Comment (Sergei-Lebedev) [inline test/gtest/core/test_allreduce.cc:265]: ifdef isn't needed here, we don't use any cuda functions

**Review by Sergei-Lebedev — APPROVED** [review]

File: src/components/tl/ucp/allreduce/allreduce_sra_knomial.c [hunk 4]
```diff
@@ -87,32 +135,74 @@ ucc_tl_ucp_allreduce_sra_knomial_init(ucc_base_coll_args_t *coll_args,
...
+                     ucc_dt_size(coll_args->args.src.info.datatype);
+    *n_frags = 1;
+    if (msgsize > cfg->allreduce_sra_kn_frag_thresh) {
+        int min_num_frags = msgsize / cfg->allreduce_sra_kn_frag_size;
+        *n_frags = ucc_max(min_num_frags, cfg->allreduce_sra_kn_n_frags);
+    }
+    *pipeline_depth = ucc_min(*n_frags, cfg->allreduce_sra_kn_pipeline_depth);
+}
+
+ucc_status_t ucc_tl_ucp_allreduce_sra_knomial_start(ucc_coll_task_t *task)
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/tl/ucp/allreduce/allreduce_sra_knomial.c:176]: i renamed finalize to ucc_tl_ucp_allreduce_sra_knomial_finalize. "knomial" better represents what the algorithm is. the fact that i can be pipelined does not matter imho. sra can be implemented with ring e.g.

File: src/schedule/ucc_schedule_pipelined.h [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+ */
+#ifndef UCC_SCHEDULE_PIPELINED_H_
+#define UCC_SCHEDULE_PIPELINED_H_
+#include "components/base/ucc_base_iface.h"
+
+#define UCC_SCHEDULE_FRAG_MAX_TASKS 8
+
+typedef struct ucc_schedule_pipelined ucc_schedule_pipelined_t;
+
+#define UCC_SCHEDULE_PIPELINED_MAX_FRAGS 4
```
Comment (bureddy) [inline src/schedule/ucc_schedule_pipelined.h:13]: question: at UCC level.  it should it make sense to use term "FRAG",  is this works only for fragmnetation (or) can we be used for any generic task pipeline?

Comment (vspetrov) [reply src/schedule/ucc_schedule_pipelined.h:13]: I'm not sure what do you mean? What is pipelining if you don't fragment the collective?

File: src/components/tl/ucp/allreduce/allreduce_sra_knomial.c [hunk 6]
```diff
@@ -87,32 +122,93 @@ ucc_tl_ucp_allreduce_sra_knomial_init(ucc_base_coll_args_t *coll_args,
...
+                                   ucc_tl_ucp_team_t *team, int *n_frags,
+                                   int *pipeline_depth)
+{
+    //TODO make selection mem_type - specific
+    ucc_tl_ucp_lib_config_t *cfg     = &UCC_TL_UCP_TEAM_LIB(team)->cfg;
+    size_t                   msgsize = coll_args->args.src.info.count *
+                     ucc_dt_size(coll_args->args.src.info.datatype);
+    *n_frags = 1;
+    if (msgsize > cfg->allreduce_sra_kn_frag_thresh) {
+        int min_num_frags = msgsize / cfg->allreduce_sra_kn_frag_size;
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/tl/ucp/allreduce/allreduce_sra_knomial.c:157]: you are right

File: src/components/tl/ucp/tl_ucp.c [hunk 7]
```diff
@@ -52,6 +52,35 @@ static ucc_config_field_t ucc_tl_ucp_lib_config_table[] = {
...
+
+    {"ALLREDUCE_SRA_KN_PIPELINE_DEPTH", "2",
+     "Number of fragments simultaneously progressed by the SRA knomial alg",
+     ucc_offsetof(ucc_tl_ucp_lib_config_t, allreduce_sra_kn_pipeline_depth),
+     UCC_CONFIG_TYPE_UINT},
+
+    {"ALLREDUCE_SRA_KN_SEQUENTIAL", "0",
+     "Type of pipelined schedule for SRA knomial alg (sequential/parallel)",
+     ucc_offsetof(ucc_tl_ucp_lib_config_t, allreduce_sra_kn_seq),
+     UCC_CONFIG_TYPE_UINT},
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/tl/ucp/tl_ucp.c:82]: agree. changed to UCC_CONFIG_TYPE_BOOL which taskes "yes/no".

**Review by bureddy — APPROVED** [review]

**Pull request merged**